include/robot_interfaces/robot_backend.hpp

```diff
@@ -26,12 +26,113 @@
 
 namespace robot_interfaces
 {
+/**
+ * @brief Timer to measure code execution time with "checkpoints"
+ *
+ * This timer is meant to be used for measuring execution time of a loop.  It
+ * measures time between calls of the `start` method, so by calling this at the
+ * beginning of the loop, you get the execution time of the full iteration.
+ * Further, you can define "checkpoints" within the loop to measure time of
+ * separate steps in the loop.  Call the `checkpoint` method after the code that
+ * is associated with it.  For each checkpoint, the time elapsed since the last
+ * checkpoint is measured (`start` counts as a checkpoint in this regard).
+ *
+ * Example:
+ *
+ * ~~~{.cpp}
+ * CheckpointTimer<3> timer;
+ *
+ * while (condition)
+ * {
+ *     timer.start();
+ *
+ *     initialize()
+ *     timer.checkpoint("initialize");
+ *
+ *     do();
+ *     some();
+ *     stuff();
+ *     timer.checkpoint("do some stuff");
+ *
+ *     write_log();
+ *     timer.checkpoint("logging");
+ * }
+ * ~~~
+ *
+ * @tparam NUM_CHECKPOINTS Number of checkpoints.
+ * @tparam ENABLED Set to false, to disable timer.  Method calls will have no
+ * effect (and should hopefully be optimized away by the compiler).
+ */
+template <size_t NUM_CHECKPOINTS, bool ENABLED = true>
+class CheckpointTimer
+{
+public:
+    //! @brief Start timer iteration.
+    void start()
+    {
+        if (ENABLED)
+        {
+            timers_[0].tac_tic();
+            current_checkpoint = 1;
+            timers_[current_checkpoint].tic();
+        }
+    }
+
+    /**
+     * @brief Set checkpoint for time measurement.
+     *
+     * Measures time from the last call of start() or checkpoint() until this
+     * call.  The given name is used when printing the results.
+     *
+     * @param checkpoint_name Name of the checkpoint (used for printing results)
+     */
+    void checkpoint(const std::string &checkpoint_name)
+    {
+        if (ENABLED)
+        {
+            timers_[current_checkpoint].tac();
+
+            if (checkpoint_names_[current_checkpoint].empty())
+            {
+                checkpoint_names_[current_checkpoint] = checkpoint_name;
+            }
+
+            current_checkpoint++;
+            if (current_checkpoint < timers_.size())
+            {
+                timers_[current_checkpoint].tic();
+            }
+        }
+    }
+
+    //! @brief Print results of time measurements.
+    void print_statistics() const
+    {
+        if (ENABLED)
+        {
+            std::cout << "===== Total:" << std::endl;
+            timers_[0].print_statistics();
+
+            for (size_t i = 1; i < timers_.size(); i++)
+            {
+                std::cout << "----- " << checkpoint_names_[i] << std::endl;
+                timers_[i].print_statistics();
+            }
+        }
+    }
+
+private:
+    std::array<real_time_tools::Timer, NUM_CHECKPOINTS + 1> timers_;
+    std::array<std::string, NUM_CHECKPOINTS + 1> checkpoint_names_;
```

Comment:
i think checkpoint_names_[0] is never used. how about setting the name to "total" ? then you also don't need to make a special case in print_statistics()

Reply:
:+1: